commons/src/scanner/symbol/iterator/mod.rs

```diff
@@ -34,6 +34,13 @@ pub struct SymbolIterator<'input> {
     end_match: Option<IteratorEndFn>,
     /// Flag set to `true` if this iterator reached its end.
     iter_end: bool,
+    /// Flag set to `true` if prefix mismatch occured.
+    prefix_mismatch: bool,
+    /// Flag set to `true` to indicate matching context in [`Self::next()`].
+    next_matching: bool,
+    /// Flag set to `true` to indicate matching context in [`Self::peeking_next()`]
+    /// Used to prevent consumed matching while peeking.
+    peek_matching: bool,
```

Comment:
I decided to have this info stored as state, because I don't see an easy way to handle this per function. The state option makes it easier with nesting iterators, because it does not affect parent iterators.
Because this state handling must only be taken into account in the iterator and matcher implementation, but not in any usage of the iterator, I think it is acceptable to have this complexity here. I added tests to ensure the behavior of peek() and next() are the same, and I don't assume that the implementation will change much in the future.

Reply:
I agree that it will probably stay like it is and won't change much. But at the same time, it is not easy to understand, and we probably will need to understand it again from time to time. So if we can improve that aspect of the code I would like to push it a little bit 😛. 

It might be the case that it's not really possible to improve that, since it's not that easy to solve anyway. But I think it's worth a try in any case 👀. 

TLDR: Complexity is (of course) acceptable, but if it's possible to reduce it then we should try anyway 👀.

Reply:
As internally discussed, we will move this optimization into its own issue.
Changes would be inside the symbol iterator, so merging this PR first helps to speed up development of other features.